### PR TITLE
feat: add warning to set functions when value is not a string (#230)

### PR DESCRIPTION
**Pull request by jkossek**
Summary:
---------
Added warning on iOS when the value in the second argument of the `setItem` method is not a string.

Test Plan:
----------
Change the second argument of the `setItem` method in `example/src/examples/GetSetClear.js`, line 41 to non-string type and then try to increase the stored value. The warning will appear.

Example:
```
await AsyncStorage.setItem(STORAGE_KEY, newNumber);
await AsyncStorage.setItem(STORAGE_KEY, {newNumber});
await AsyncStorage.setItem(STORAGE_KEY, [newNumber]);
await AsyncStorage.setItem(STORAGE_KEY, null);
```

### CODE REVIEW

File: lib/AsyncStorage.js [hunk 1]
```diff
@@ -103,7 +103,19 @@ const AsyncStorage = {
         const errs = convertErrors(errors);
         callback && callback(errs && errs[0]);
         if (errs) {
-          reject(errs[0]);
+          if (Platform.OS !== 'ios') {
+            reject(errs[0]);
+          } else if (
+            errs[0] &&
+            String(errs[0]).match('Entries must be arrays of the form') &&
```
Comment (Krizzu) [inline lib/AsyncStorage.js]: Turns out, the error `Entries must be arrays of the form` from native was shown when somehow wrongly parsed data was given. Since I upgraded the RN, it's no longer showing up.

We can revert those changes back to checking if value is not type of string, then displaying our warning.

Comment (jkossek) [reply lib/AsyncStorage.js]: @Krizzu I've reverted it to the previous version.

File: lib/AsyncStorage.js [hunk 2]
```diff
@@ -100,6 +100,11 @@ const AsyncStorage = {
   ): Promise<null> {
     return new Promise((resolve, reject) => {
       RCTAsyncStorage.multiSet([[key, value]], function(errors) {
+        if (Platform.OS === 'ios' && typeof value !== 'string') {
```
Comment (tido64) [inline lib/AsyncStorage.js]: Any reason why we only do this on iOS?

File: lib/AsyncStorage.js [hunk 3]
```diff
@@ -100,6 +100,11 @@ const AsyncStorage = {
...
+          console.warn(
```
Comment (tido64) [inline lib/AsyncStorage.js]: There are a lot of completely unrelated changes in this PR. Can we split them out to separate PRs? The scope of this one should only contain changes to this file.

Comment (Krizzu) [reply lib/AsyncStorage.js]: I had an issue to run the example app with Xcode 11, so I upgraded it. But good point, I'll create a different PR for that

File: lib/AsyncStorage.js (hunk 2 above)
Comment (Krizzu) [reply lib/AsyncStorage.js]: Just checked against android - this this warning should also be shown.

Comment (Krizzu) [reply lib/AsyncStorage.js]: Also: on `multiSet` should also display warning

File: lib/AsyncStorage.js [hunk 4]
```diff
@@ -301,6 +310,12 @@ const AsyncStorage = {
     callback?: ?(errors: ?$ReadOnlyArray<?Error>) => void,
   ): Promise<null> {
     return new Promise((resolve, reject) => {
+      keyValuePairs.forEach(([key, value]) => {
+        if (typeof value !== 'string') {
```
Comment (tido64) [inline lib/AsyncStorage.js]: Should we move the check inside the new function as well?

**Comment by Krizzu** [comment]
thanks @jkossek @tido64 !
